### PR TITLE
Adding CloudEvents output option.

### PR DESCRIPTION
**Pull request by n3wscott**
/kind feature
/area outputs

**What this PR does / why we need it**:

Adding CloudEvents Output support, starting with PoC of only via http binary mode.

### CODE REVIEW

File: outputs/cloudevents.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	c.Stats.CloudEvents.Add(Total, 1)
+
+	if c.ce == nil {
+		client, err := cloudevents.NewDefaultClient()
+		if err != nil {
+			go c.CountMetric(Outputs, 1, []string{"output:cloudevents", "status:error"})
+			log.Println("[ERROR] : CloudEvents - NewDefaultClient", err)
+			return
+		}
+		c.ce = client
```
Comment (Issif) [inline outputs/cloudevents.go]:
```suggestion
		c.CloudEventsClient = &client
```

File: outputs/cloudevents.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	ctx := cloudevents.ContextWithTarget(context.Background(), c.EndpointURL.String())
+
+	event := cloudevents.NewEvent()
+	event.SetTime(falcopayload.Time)
+	event.SetSource("falco.org") // TODO: this should have some info on the falco server that made the event.
```
Comment (Issif) [inline outputs/cloudevents.go:29]:
```suggestion
	event.SetSource("falco") // TODO: this should have some info on the falco server that made the event.
```

`falco` doesn't send this intel (`host` or else) in its output.

File: outputs/cloudevents.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	go c.CountMetric(Outputs, 1, []string{"output:cloudevents", "status:ok"})
+	c.Stats.Webhook.Add(OK, 1)
+	c.PromStats.Outputs.With(map[string]string{"destination": "cloudevents", "status": OK}).Inc()
+	log.Println("[INFO] : CloudEvents - Send OK")
+}
+
+func toEventType(rule string) string {
+	rule = strings.ToLower(rule)
+
+	// TODO: this could be improved but I do not know what is possible, I am assuming it is a free form entry.
```
Comment (Issif) [inline outputs/cloudevents.go]: yes it is

File: outputs/cloudevents.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	for _, old := range []string{".", ":", ","} {
```
Comment (Issif) [inline outputs/cloudevents.go]: You may reuse this pattern : https://github.com/falcosecurity/falcosidekick/blob/0403fbd025724f4eb9ea273fbc5f06a277296b23/outputs/nats.go#L28

File: outputs/cloudevents.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	event.SetExtension("priority", falcopayload.Priority)
+	event.SetExtension("rule", falcopayload.Rule)
+
+	if err := event.SetData(cloudevents.ApplicationJSON, falcopayload); err != nil {
+		log.Println("[ERROR] : CloudEvents, failed to set data", err)
+	}
+
+	if result := c.ce.Send(ctx, event); cloudevents.IsUndelivered(result) {
+		go c.CountMetric(Outputs, 1, []string{"output:cloudevents", "status:error"})
+		c.Stats.Webhook.Add(Error, 1)
```
Comment (Issif) [inline outputs/cloudevents.go:45]:
```suggestion
		c.Stats.CloudEvents.Add(Error, 1)
```

File: outputs/cloudevents.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+	event := cloudevents.NewEvent()
+	event.SetTime(falcopayload.Time)
+	event.SetSource("falco.org") // TODO: this should have some info on the falco server that made the event.
+	event.SetType(toEventType(falcopayload.Rule))
+	event.SetExtension("priority", falcopayload.Priority)
+	event.SetExtension("rule", falcopayload.Rule)
+
+	if err := event.SetData(cloudevents.ApplicationJSON, falcopayload); err != nil {
+		log.Println("[ERROR] : CloudEvents, failed to set data", err)
```
Comment (Issif) [inline outputs/cloudevents.go:40]:
```suggestion
		log.Printf("[ERROR] : CloudEvents, failed to set data : %v\n", err)
```

File: outputs/client.go [hunk 7]
```diff
@@ -44,17 +45,18 @@ var ErrClientCreation = errors.New("Client creation Error")
...
+	Config            *types.Configuration
+	Stats             *types.Statistics
+	PromStats         *types.PromStatistics
+	AWSSession        *session.Session
+	StatsdClient      *statsd.Client
+	DogstatsdClient   *statsd.Client
+	GCPTopicClient    *pubsub.Topic
+	KafkaProducer     *kafka.Conn
+	PagerdutyClient   *pagerduty.Client
+	CloudEventsClient cloudevents.Client
```
Comment (Issif) [inline outputs/client.go:59]:
```suggestion
	CloudEventsClient *cloudevents.Client
```

File: outputs/cloudevents.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	c.Stats.CloudEvents.Add(Total, 1)
+
+	if c.CloudEventsClient == nil {
+		client, err := cloudevents.NewDefaultClient()
+		if err != nil {
+			go c.CountMetric(Outputs, 1, []string{"output:cloudevents", "status:error"})
+			log.Println("[ERROR] : CloudEvents - NewDefaultClient", err)
+			return
+		}
+		c.CloudEventsClient = client
```
Comment (Issif) [inline outputs/cloudevents.go:22]:
```suggestion
		c.CloudEventsClient = &client
```

File: README.md [hunk 9]
```diff
@@ -377,6 +377,9 @@ The *env vars* "match" field names in *yaml file with this structure (**take car
 * **WEBHOOK_ADDRESS** : "" # Webhook address, if not empty, Webhook output is *enabled*
 * **WEBHOOK_CUSTOMHEADERS** : a list of comma separated custom headers to add, syntax is "key:value,key:value"
 * **WEBHOOK_MINIMUMPRIORITY** : minimum priority of event for using this output, order is `emergency|alert|critical|error|warning|notice|informational|debug or "" (default)`
+* **CLOUDEVENTS_ADDRESS** : "" # Webhook address, if not empty, Webhook output is *enabled*
```
Comment (Issif) [inline README.md:380]:
```suggestion
* **CLOUDEVENTS_ADDRESS** : Cloudevent address, if not empty, Cloudevents output is *enabled*
```

File: outputs/cloudevents.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+	if err := event.SetData(cloudevents.ApplicationJSON, falcopayload); err != nil {
+		log.Println("[ERROR] : CloudEvents, failed to set data", err)
+	}
+
+	if result := c.CloudEventsClient.Send(ctx, event); cloudevents.IsUndelivered(result) {
+		go c.CountMetric(Outputs, 1, []string{"output:cloudevents", "status:error"})
+		c.Stats.Webhook.Add(Error, 1)
+		c.PromStats.Outputs.With(map[string]string{"destination": "cloudevents", "status": Error}).Inc()
+		log.Println("[ERROR] : CloudEvents -", result)
```
Comment (Issif) [inline outputs/cloudevents.go:47]:
```suggestion
		log.Printf("[ERROR] : CloudEvents - %v\n", result)
```

File: outputs/cloudevents.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+		return
+	}
+
+	// Setting the success status
+	go c.CountMetric(Outputs, 1, []string{"output:cloudevents", "status:ok"})
+	c.Stats.Webhook.Add(OK, 1)
```
Comment (Issif) [inline outputs/cloudevents.go:53]:
```suggestion
	c.Stats.CloudEvents.Add(OK, 1)
```

**Review by Issif — CHANGES_REQUESTED** [review]

File: outputs/client.go (hunk 7 above)
Comment (n3wscott) [reply outputs/client.go:59]: `cloudevents.Client` is an interface.

File: outputs/cloudevents.go (hunk 8 above)
Comment (n3wscott) [reply outputs/cloudevents.go:22]: client is a pointer.

File: outputs/cloudevents.go [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+	"github.com/falcosecurity/falcosidekick/types"
+)
+
+// CloudEvent produces a CloudEvent and sends to the CloudEvents consumers.
+func (c *Client) CloudEvents(falcopayload types.FalcoPayload) {
+	c.Stats.CloudEvents.Add(Total, 1)
+
+	if c.ce == nil {
+		client, err := cloudevents.NewDefaultClient()
```
[the remark this answers is not in the document]
Comment (n3wscott) [reply outputs/cloudevents.go:16]: it is stored on line 22, it is only if it is nil.

File: outputs/cloudevents.go (hunk 2 above)
Comment (n3wscott) [reply outputs/cloudevents.go:29]: If it did, you would know which falco service has made this observation.

File: outputs/cloudevents.go (hunk 4 above)
Comment (n3wscott) [reply outputs/cloudevents.go]: I ended up deciding the event type can be the same, and the rule name is added as an extension

File: README.md [hunk 13]
```diff
@@ -377,6 +377,9 @@ The *env vars* "match" field names in *yaml file with this structure (**take car
 * **WEBHOOK_ADDRESS** : "" # Webhook address, if not empty, Webhook output is *enabled*
 * **WEBHOOK_CUSTOMHEADERS** : a list of comma separated custom headers to add, syntax is "key:value,key:value"
 * **WEBHOOK_MINIMUMPRIORITY** : minimum priority of event for using this output, order is `emergency|alert|critical|error|warning|notice|informational|debug or "" (default)`
+* **CLOUDEVENTS_ADDRESS** : "" # CloudEvents consumer address, if not empty, CloudEvents output is *enabled*
```
Comment (Issif) [inline README.md:380]:
```suggestion
* **CLOUDEVENTS_ADDRESS** : CloudEvents consumer address, if not empty, CloudEvents output is *enabled*
```

File: outputs/cloudevents.go (hunk 2 above)
Comment (Issif) [reply outputs/cloudevents.go:29]: what do you call `falco service`?

**Review by Issif — CHANGES_REQUESTED** [review]

File: outputs/cloudevents.go (hunk 2 above)
Comment (n3wscott) [reply outputs/cloudevents.go:29]: Ideally, you can find the falco install using a combo of `cloudevent.source` and `cloudevents.subject`

Comment (n3wscott) [reply outputs/cloudevents.go:29]: source is suppose to be a uri-ref...

Comment (n3wscott) [reply outputs/cloudevents.go:29]: maybe change it in a followup?

File: handlers.go [hunk 14]
```diff
@@ -210,6 +210,10 @@ func forwardEvent(falcopayload types.FalcoPayload) {
 		go webhookClient.WebhookPost(falcopayload)
 	}
 
+	if config.CloudEvents.Address != "" && (priorityMap[strings.ToLower(falcopayload.Priority)] >= priorityMap[strings.ToLower(config.CloudEvents.MinimumPriority)] || falcopayload.Rule == TestRule) {
+		go cloudeventsClient.CloudEventsPost(falcopayload)
```
Comment (KeisukeYamashita) [inline handlers.go:214]: As a suggestion, why don't we make the name of the method that fires the event strict?

We use `Publish` for GCP PubSub, `Produce` for Kafka Producer, `Post` for Webhooks, API (e.g. `SlackPost`, `WebhookPost`), etc.
In the CloudEvent SDK, `Send` is used to fire events to the backend (ref: https://github.com/cloudevents/sdk-go/blob/93f9fa1cd3103d71d1b609a75ec3160ed601b436/v2/client/client.go#L22). 

Suggestion: 

```suggestion
		go cloudeventsClient.CloudEventsSend(falcopayload)
```

If this looks good, please rename the method in `outputs/cloudevent.go` too  🙇

**Review by KeisukeYamashita — COMMENTED** [review]
Overall LGTM but I have a comment, PTAL.

File: outputs/cloudevents.go (hunk 2 above)
Comment (Issif) [reply outputs/cloudevents.go:29]: If it's supposed to be an URI, ok for `falco.org`

**Comment by Issif** [comment]
/hold

The PR seems ok for me now, but I don't want an auto-merge before `2.20.0` is released. Maybe some fields have to be discussed too.

**Comment by n3wscott** [comment]
Rebased to HEAD.

I ran the lint tool on the PR and it updated the formatting on the README.md. It uses prettier.io 's rules so let me know if you do not like that.

**Comment by Issif** [comment]
@n3wscott I checked the README, the linting seems good for me, I'm not opiniated.

**Comment by Issif** [comment]
@n3wscott can you squash the commits please?

**Comment by n3wscott** [comment]
Can you squash merge instead?

**Comment by Issif** [comment]
> Can you squash merge instead?

Merges are automatically made after an approval by poiana bot, it doesn't squash merge as far I know.

**Review by Issif — APPROVED** [review]
thank you @n3wscott 🙏

**Pull request merged**